homeassistant/components/plaato/__init__.py

```diff
@@ -60,31 +85,134 @@
+                hass.config_entries.async_forward_entry_setup(entry, platform)
+            )
 
     return True
 
 
-async def async_unload_entry(hass, entry):
+def setup_webhook(hass: HomeAssistant, entry: ConfigEntry):
+    """Init webhook based on config entry."""
+    if entry.data[CONF_WEBHOOK_ID] is None:
+        raise InvalidStateError
+
+    webhook_id = entry.data[CONF_WEBHOOK_ID]
+    device_name = entry.data[CONF_DEVICE_NAME]
+
+    __set_entry_data(entry, hass)
+
+    hass.components.webhook.async_register(
+        DOMAIN, f"{DOMAIN}.{device_name}", webhook_id, handle_webhook
+    )
+
+
+async def async_setup_coordinator(hass: HomeAssistant, entry: ConfigEntry):
+    """Init auth token based on config entry."""
+    auth_token = entry.data[CONF_TOKEN]
+    device_type = entry.data[CONF_DEVICE_TYPE]
+
+    if entry.options.get(CONF_SCAN_INTERVAL):
+        update_interval = timedelta(minutes=entry.options[CONF_SCAN_INTERVAL])
+    else:
+        update_interval = DEFAULT_SCAN_INTERVAL
+
+    coordinator = PlaatoCoordinator(hass, auth_token, device_type, update_interval)
+    await coordinator.async_refresh()
+    if not coordinator.last_update_success:
+        raise ConfigEntryNotReady
+
+    __set_entry_data(entry, hass, coordinator, auth_token)
+
+    for platform in PLATFORMS:
+        if entry.options.get(platform, True):
+            coordinator.platforms.append(platform)
+
+
+def __set_entry_data(entry, hass, coordinator=None, device_id=None):
+    device = {
+        DEVICE_NAME: entry.data[CONF_DEVICE_NAME],
+        DEVICE_TYPE: entry.data[CONF_DEVICE_TYPE],
+        DEVICE_ID: device_id,
+    }
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        COORDINATOR: coordinator,
+        DEVICE: device,
+        SENSOR_DATA: None,
+        UNDO_UPDATE_LISTENER: entry.add_update_listener(_async_update_listener),
+    }
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
     """Unload a config entry."""
-    hass.components.webhook.async_unregister(entry.data[CONF_WEBHOOK_ID])
-    hass.data[SENSOR_DATA_KEY]()
+    use_webhook = entry.data[CONF_USE_WEBHOOK]
+    hass.data[DOMAIN][entry.entry_id][UNDO_UPDATE_LISTENER]()
 
-    await hass.config_entries.async_forward_entry_unload(entry, SENSOR)
-    return True
+    if use_webhook:
+        return await async_unload_webhook(hass, entry)
+
+    return await async_unload_coordinator(hass, entry)
+
+
+async def async_unload_webhook(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload webhook based entry."""
+    if entry.data[CONF_WEBHOOK_ID] is not None:
+        hass.components.webhook.async_unregister(entry.data[CONF_WEBHOOK_ID])
+
+    unloaded = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, platform)
+                for platform in PLATFORMS
+            ]
+        )
+    )
+    if unloaded:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unloaded
```

Comment:
This can be broken out into a new function that accepts a list of platforms to unload.

You could pass `PLATFORMS` from here and

`coordinator.platforms` in the other function